File: src/main/java/at/xoola/heartdrops/DropTask.java
package at.xoola.heartdrops;

import org.bukkit.Location;
import org.bukkit.Particle;
import org.bukkit.entity.Player;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

class DropTask implements Runnable {
    private final Map<Integer, Location> drops = new ConcurrentHashMap<>();

    public void addDrop(int entityId, Location loc) {
        if (loc != null && loc.getWorld() != null) {
            drops.put(entityId, loc.clone());
        }
    }

    public void clearDrops() {
        HeartDisplayManager manager = HeartDrops.instance.getHeartDisplayManager();
        if (manager != null) {
            manager.clearAll();
        }
        drops.clear();
    }

    @Override
    public void run() {
        HeartDisplayManager manager = HeartDrops.instance.getHeartDisplayManager();
        Iterator<Map.Entry<Integer, Location>> it = drops.entrySet().iterator();
        
        while (it.hasNext()) {
            Map.Entry<Integer, Location> entry = it.next();
            int entityId = entry.getKey();
            Location loc = entry.getValue();
            
            if (loc.getWorld() == null) {
                if (manager != null) {
                    manager.removeHeart(entityId);
                }
                it.remove();
                continue;
            }

            // Spawn particles around the heart
            loc.getWorld().spawnParticle(Particle.HEART, loc, 3, 0.4, 0.4, 0.4, 0.0);

            // Check for nearby players
            for (Player p : loc.getWorld().getPlayers()) {
                if (!p.hasPermission("heartdrops.pickup")) continue;
                
                if (p.getLocation().distanceSquared(loc) < 2.25) {
                    HeartDrops.instance.healPlayer(p);
                    
                    if (manager != null) {
                        manager.removeHeart(entityId);
                    }
                    it.remove();
                    break;
                }
            }
        }
    }
}

File: src/main/kotlin/at/xoola/heartdrops/DropTask.kt
package at.xoola.heartdrops

import org.bukkit.Location
import org.bukkit.Particle
import java.util.concurrent.ConcurrentHashMap

internal class DropTask : Runnable {
    private val drops = ConcurrentHashMap<Int, Location>()

    fun addDrop(entityId: Int, loc: Location?) {
        if (loc != null && loc.world != null) {
            drops[entityId] = loc.clone()
        }
    }

    fun clearDrops() {
        val manager = HeartDrops.instance?.heartDisplayManager
        manager?.clearAll()
        drops.clear()
    }

    override fun run() {
        val manager = HeartDrops.instance?.heartDisplayManager
        val iterator = drops.entries.iterator()
        
        while (iterator.hasNext()) {
            val (entityId, loc) = iterator.next()
            
            if (loc.world == null) {
                manager?.removeHeart(entityId)
                iterator.remove()
                continue
            }

            // Spawn particles around the heart
            loc.world?.spawnParticle(Particle.HEART, loc, 3, 0.4, 0.4, 0.4, 0.0)

            // Check for nearby players
            for (player in loc.world?.players ?: emptyList()) {
                if (!player.hasPermission("heartdrops.pickup")) continue
                
                if (player.location.distanceSquared(loc) < 2.25) {
                    HeartDrops.instance?.healPlayer(player)
                    manager?.removeHeart(entityId)
                    iterator.remove()
                    break
                }
            }
        }
    }
}
